### PR TITLE
Ntp offset

### PR DESCRIPTION
**Pull request by chbatey**

### CODE REVIEW

File: ntp.go [hunk 1]
```diff
@@ -62,52 +80,81 @@ func (m *msg) SetMode(md mode) {
...
+	return offset(m.OriginTime, m.ReceiveTime, m.TransmitTime, toNtpTime(time.Now()))
+}
+
+// offset variable names based off the rfc:
+// https://tools.ietf.org/html/rfc2030
+func offset(t1, t2, t3, t4 ntpTime) (int64, error) {
+	// https://tools.ietf.org/html/rfc2030 page 12
+	// d = (T4 - T1) - (T2 - T3)
+	// t = ((T2 - T1) + (T3 - T4)) / 2
+	i := float64((t2.nsec() - t1.nsec()) + (t3.nsec() - t4.nsec()))
```
Comment (jsravn) [inline ntp.go]:
```
(t2.nsec - t1.nsec / 2) + (t3.nsec - t4.nsec / 2)
```

Should help prevent overflows.

File: ntp.go [hunk 2]
```diff
@@ -62,52 +80,81 @@ func (m *msg) SetMode(md mode) {
...
+func Offset(host string) (int64, error) {
+	m, err := getTime(host, 4)
+	if err != nil {
+		return 0, err
+	}
+	return offset(m.OriginTime, m.ReceiveTime, m.TransmitTime, toNtpTime(time.Now()))
+}
+
+// offset variable names based off the rfc:
+// https://tools.ietf.org/html/rfc2030
```
Comment (jsravn) [inline ntp.go]: What about just copying the table from the rfc?
```
      Timestamp Name          ID   When Generated
      ------------------------------------------------------------
      Originate Timestamp     T1   time request sent by client
      Receive Timestamp       T2   time request received by server
      Transmit Timestamp      T3   time reply sent by server
      Destination Timestamp   T4   time reply received by client
```

File: ntp.go [hunk 3]
```diff
@@ -62,52 +80,81 @@ func (m *msg) SetMode(md mode) {
...
-		return time.Now(), err
+		return nil, err
 	}
 	defer con.Close()
 	con.SetDeadline(time.Now().Add(5 * time.Second))
 
 	m := new(msg)
 	m.SetMode(client)
 	m.SetVersion(version)
+	m.SetTransmitTime(time.Now())
```
Comment (jsravn) [inline ntp.go]: Isn't transmit time set by the server? Client should set the origin timestamp.

Comment (chbatey) [inline ntp.go]:
```
 To calculate the roundtrip delay d and local clock offset t relative
   to the server, the client sets the transmit timestamp in the request
   to the time of day according to the client clock in NTP timestamp
   format. The server copies this field to the originate timestamp in
   the reply and sets the receive timestamp and transmit timestamp to
   the time of day according to the server clock in NTP timestamp
   format.
```

Comment (chbatey) [inline ntp.go]: The client sets the transmit tine on the way out and the server copies it across

File: ntp.go (hunk 2 above)
Comment (chbatey) [inline ntp.go]: boo I already renamed them all :)

File: ntp.go (hunk 3 above)
Comment (jsravn) [inline ntp.go]: Ok got it.

**Pull request merged**